### PR TITLE
Detect multiprotocol keywords 7304 v1

### PR DESCRIPTION
**Pull request by catenacyber**
Link to ticket: https://redmine.openinfosecfoundation.org/issues/
https://redmine.openinfosecfoundation.org/issues/7304

Describe changes:
- cleaner code for multi app-layer keywords such as ja4

Is there something to do with DCERPC/SMB stuff ?

I think file keywords have their own logic and it is ok.

### CODE REVIEW

File: src/detect-ja4-hash.c [hunk 1]
```diff
@@ -174,4 +180,64 @@ static InspectionBuffer *Ja4DetectGetHash(DetectEngineThreadCtx *det_ctx,
...
+                        "\"q13d0310h3_55b375c5d22e_cd85d2d88918\";)");
+    // cannot have a smb keyword with ja4.hash (quic or tls)
+    FAIL_IF_NOT_NULL(s);
+
+    // valid tests
+    s = SigInit(de_ctx, "alert ip any any -> any any (sid: 1; "
+                        "ja4.hash; content: \"q13d0310h3_55b375c5d22e_cd85d2d88918\";)");
+    // just ja4.hash any proto
+    FAIL_IF_NULL(s);
+    s = SigInit(de_ctx, "alert quic any any -> any any (sid: 1; "
```
Comment (victorjulien) [inline src/detect-ja4-hash.c:220]: this adds another sid:1, is that not rejected as a duplicate? Or do we only check later? Might be good to use unique id's for all sigs, but esp for the good ones

File: src/detect-parse.c [hunk 2]
```diff
@@ -1734,6 +1734,80 @@ int DetectSignatureAddTransform(Signature *s, int transform, void *options)
     SCReturnInt(0);
 }
 
+int DetectSignatureSetMultiAppProto(Signature *s, const AppProto *alprotos)
```
Comment (victorjulien) [inline src/detect-parse.c:1737]: should add comment about how alprotos is an array terminated by `ALPROTO_UNKNOWN`

**Comment by victorjulien** [comment]
I'm missing a lot of explanation on the new logic.

**Comment by catenacyber** [comment]
> I'm missing a lot of explanation on the new logic.

Motivation is to have a cleaner support for multi-protocol keywords such as JA3 both on QUIC and TLS.
We do not want hardcoding, ie we do not to see `ALPROTO_QUIC` constant directly used in generic src/detect-parse.c

From the keyword point of view, I propose to add `DetectSignatureSetMultiAppProto` that is like `DetectSignatureSetAppProto` but takes multiple alprotos, but behaves in the same way, doing intersection and rejecting a signature early if there is a app-layer incompatibility...

From the data structure, I propose to add a fixed-length array in SignatureInitData, as the use case is a sparse number of protocols, not taking half of the 32 protocols we support...

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 23609

File: src/detect-parse.c (hunk 2 above)
Comment (catenacyber) [reply src/detect-parse.c:1737]: right

**Comment by catenacyber** [comment]
Next in https://github.com/OISF/suricata/pull/12229

**Pull request closed without merging**